src/mo_frontend/parsing.ml

```diff
@@ -70,39 +70,95 @@ let slice_lexeme lexbuf i1 i2 =
   then "<unknown>" (* Too rare to care *)
   else Bytes.sub_string lexbuf.lex_buffer offset len
 
-let parse mode error_detail checkpoint lexer lexbuf =
-  Diag.with_message_store (fun m ->
-    try
-      (* Temporary hack! *)
-      Parser_lib.msg_store := Some m;
-      Parser_lib.mode := Some mode;
-      Some (E.entry checkpoint lexer)
-    with E.Error ((start, end_), explanations) ->
-      let at =
+module I = Parser.MenhirInterpreter
+
+(* For debug *)
+(* module RecoveryTracer = MenhirRecoveryLib.MakePrinter ( *)
+(*   struct *)
+(*     module I = Parser.MenhirInterpreter *)
+(*     let print s = Printf.eprintf "%s" s *)
+(*     let print_symbol s = print (Printers.string_of_symbol s) *)
+(*     let print_element = None *)
+(*     let print_token t = print (Source_token.string_of_parser_token t) *)
+(*   end) *)
+
+module RecoveryTracer = MenhirRecoveryLib.DummyPrinter (I)
+
+module RecoveryConfig = struct
+  include Recover_parser
+
+  (* Adapt [default_value region] to MenhirRecoverLib interface ([default_value loc]) *)
+  let default_value (loc: Custom_compiler_libs.Location.t) sym =
+      let open Custom_compiler_libs.Location in
+      let open Lexing in
+      let open Source in
+      let file = loc.loc_start.pos_fname in
+      let region_loc : region = {
+        left  : pos = {file; line = loc.loc_start.pos_lnum; column = loc.loc_start.pos_bol};
+        right : pos = {file; line = loc.loc_end.pos_lnum; column = loc.loc_end.pos_bol};
+      } in
+      default_value region_loc sym
```

Comment:
Is this `default_value` obtained from one of the open modules above? Otherwise I looks non-terminating...

Reply:
This is a bit awkward part of the recovery API where we tried to allow to have any location type.

So `default_value` is generated in `recover_parser.ml` and doesn't restrict location type, so you can use any location type in the `[@recover.expr ..]` attributes of the grammar. This `default_value` will have `Source.region` -> 'a I.symbol -> 'a` type. 

But to fit the signature of `MenhirRecoveryLib.Make`, you have to provide `default_value` of type `Custom_compiler_libs.Location.t -> 'a I.symbol -> 'a`. So we define `default_value` with the right type via the generated converting location type and calling the generated `default_value :: Source.region -> ..`

Reply:
So those `default_value` are actually different functions, no recursion